Fixed Ubercharge ending too early on Medics

Starblaster64 opened the pull request:
Uses hidden CVAR that controls Uber drain rate to fix the issue of Ubercharge ending too early on Medics. Lasts for 12 second as before, however the actual drain rate has been reduced by 33%.
Uber meter is no longer altered on Uber activation. 
Third Degree can grant Ubercharge only up to 100% instead of 150% as a result.

EDIT: Probably going to squash all these commits before the merge is accepted (if requested), would be rather messy otherwise.

Starblaster64 committed:
Re-worked Medigun Ubercharge

No longer sets Ubercharge to 150% on an Uber. Mediguns have an attribute that makes Uber last 4 seconds longer, reducing the Uber drain rate by 33%. Third Degree maximum Uber threshold is reduced to 100% from 150%. Fixes the issue with Ubercharges ending early on Medics who pop them.

Starblaster64 committed:
Re-reworked Ubercharge

The Uber duration attribute didn't fix the problem for some reason. Switched over to using the hidden Medigun Charge Release Rate CVAR.

Chdata commented:
Setting max to 1.0 if m_bChargeReleased is true is pointless, you can remove the max variable altogether. (I'd comment that stuff out).

Starblaster64 commented:
What, the Third Degree thing? No it isn't, stops TD getting Ubercharge over 100% while it's draining. With the Uber drain rate change, you'll gain 3.33% Uber per hit. Nothing else in the game lets you obtain over 100% Ubercharge, and with this change nothing in VSH will either.

EDIT: I think I see what you meant. You want to remove the part that sets the max var during Ubercharge, since it's defined beforehand already. And then remove that too since it's a static number.

Starblaster64 committed:
Removed redundant Ubercharge meter check

Chdata commented:
yes the static number

Starblaster64 commented:
Put this PR on the back burner for the moment. 
Someone pointed out on Reddit that Medic-chains are affected by this bug too. Which is to say, if we have two Medics (A & B), and A pops Uber on B, then B pops their Uber halfway through, if A stops healing B or simply waits until their Uber runs out, B's Uber will be cancelled out along with it.

This PR change wouldn't solve that. Also, now that it has been posted on Reddit there's a chance that Valve will give some priority to the issue, and we may see it patched in a future update (hopefully soon).

Chdata commented:
lol

shadow93 commented:
Nah, lets put it on the degreaser.

Starblaster64 commented:
OK, after the Tough Break update it's still broken. Valve has allegedly fixed the issue in MvM at least, so I'm going to attempt using the attribute again, since that would present itself more clearly to players than modifying the CVAR would.

EDIT: Nope, probably only fixed it while the gamemode logic is MvM or whatever.

Starblaster64 commented:
Here's an idea, what if we leave everything as it is, and just add a ClientTimer check onto the Medic?

**To elaborate:** A Medic who has their Medigun out, and Ubercharge draining above 0% will have the Uber condition constantly applied to them for an Infinite duration. When their Ubercharge reaches 0%, the game will take over and remove condition 5 after 1 second, and add the Uber flicker condition during that time.
Weapon switch wouldn't need to be considered, since switching weapons would 'reset' the Uber automatic removal thing, _and_, upon switching to another weapon the game automatically removes condition 5 (Uber) from the Medic.
The heal target wouldn't need to be considered, since they are unaffected by the bug (with the exception of a very small interruption of Uber, only a few frames).

EDIT: Tested this method and found no issues so far.

Starblaster64 committed:
Re-re-re-worked Ubercharge

Uses ClientTimer to add the Uber condition to Medics who have their Medigun out and draining with above 0% Ubercharge.
Game logic handles the rest.

Chdata commented:
It's interesting that that worked.

Starblaster64 commented:
Why's that?

Chdata commented:
i forget

shadow93 commented:
frog

Chdata commented:
i froget

Starblaster64 committed:
Converted tabs to spaces
